hail/src/main/scala/is/hail/expr/ir/FreeVariables.scala

```diff
@@ -65,26 +138,48 @@ object FreeVariables {
                 val bindings = Bindings.segregated(ir1, i, baseEnv)
                 val childEnv = bindings.childEnvWithoutBindings
                 val sub = compute(child, childEnv).subtract(bindings.newBindings)
-                if (UsesAggEnv(ir1, i))
+                val env = if (UsesAggEnv(ir1, i))
                   sub.copy(eval = Env.empty[Unit], agg = Some(sub.eval), scan = baseEnv.scan)
                 else if (UsesScanEnv(ir1, i))
                   sub.copy(eval = Env.empty[Unit], agg = baseEnv.agg, scan = Some(sub.eval))
                 else
                   sub
+                assert(
+                  (env.agg.isDefined == baseEnv.agg.isDefined || env.scan.isDefined == baseEnv.scan.isDefined)
+                )
+                env
               case _ =>
                 baseEnv
             }
             .fold(baseEnv)(_.merge(_))
       }
     }
 
-    compute(
+    val old = compute(
       ir,
       BindingEnv(
         Env.empty,
         if (supportsAgg) Some(Env.empty[Unit]) else None,
         if (supportsScan) Some(Env.empty[Unit]) else None,
       ),
     )
+
+    val new_ = {
+      val env = FreeVariableBindingEnv(
+        Some(new FreeVariableEnv(Env.empty)),
+        if (supportsAgg) Some(new FreeVariableEnv(Env.empty)) else None,
+        if (supportsScan) Some(new FreeVariableEnv(Env.empty)) else None,
+      )
+      VisitIR.withEnv(ir, env) { (ir, env) =>
+        ir match {
+          case Ref(name, _) => env.visitRef(name)
+          case _ =>
+        }
+      }
+      env.getFreeVars
+    }
+
+    assert(old == new_, s"old: $old\nnew: ${new_}\nir:\n${Pretty.sexprStyle(ir, allowUnboundRefs = true)}")
```

Comment:
Note that all tests passed with this assertion, checking that the new free variables implementation always agrees with the old.